tools/seq-data-to-lane-bam/seq-data-to-lane-bam.py

```
@@ -71,10 +71,15 @@ def group_readgroup_by_filepair(seq_experiment_analysis):
     return filepair_map_to_readgroup
 
 
-def readgroup_id_to_fname(rg_id):
+def readgroup_id_to_fname(rg_id, study_id=None, donor_id=None, sample_id=None):
     friendly_fname = "".join([ c if re.match(r"[a-zA-Z0-9\.\-_]", c) else "_" for c in rg_id ])
     md5sum = hashlib.md5(rg_id.encode('utf-8')).hexdigest()
-    return "%s.%s.lane.bam" % (friendly_fname, md5sum)
+    fname_parts = [friendly_fname, md5sum, 'lane.bam']
+    if sample_id: fname_parts.insert(0, sample_id)
+    if donor_id: fname_parts.insert(0, donor_id)
+    if study_id: fname_parts.insert(0, study_id)
```

Do we allow these three ids not be provided?

---

in our case, no, they will never be `None`. When we get the values for these we can them from sequencing experiment metadata like below:
```
    study_id = seq_experiment_analysis['studyId']
    donor_id = seq_experiment_analysis['samples'][0]['donor']['donorId']
    sample_id = seq_experiment_analysis['samples'][0]['sampleId']
```
We know that these are all mandatory, so none of the IDs will be None.

---

we can error out if any of these is missing, that should be safer

---

Exactly, that's what I understand.

---

I think error out is bette, since we do not allow any of those three ids will be missing. If indeed they are missing for any reason, we should error out instead of silent accepting.